### ISSUE TITLE
Clarify that nan is supported in zero_division

### ISSUE THREAD

**Issue by robmarkcole** [Lightning-AI/torchmetrics#2535]
This option added in https://github.com/Lightning-AI/torchmetrics/pull/2198 to create equivalence with [sklearn](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.precision_score.html#sklearn-metrics-precision-score) also supports passing through a nan if I am not mistaken, based on the implementation of `torchmetrics.utilities.compute._safe_divide`.
I wanted to clarify that nan can indeed be passed through, ask if any code changed are required to support this officially, and suggest this could be added to the docs which state currently `zero_division [float] – Should be 0 or 1.`, e.g. [here](https://lightning.ai/docs/torchmetrics/v1.4.0/classification/precision.html#binary-precision)

**Comment by SkafteNicki** [comment]
Hi @robmarkcole, thanks for raising this issue and sorry for the slow reply from my side.
So currently `nan` is not a supported value, only 0 and 1 are supported. I can try to look into what it will take to support `nan` as a value. To my understanding the `nan` option means in sklearn that values will be ignored when calculating the average. I can try to look into what it will take to implement this into torchmetrics

**Comment by robmarkcole** [comment]
@SkafteNicki thanks for clarifying, and yes ignoring these values in averages is the goal 👍

**Comment by DimitrisMantas** [comment]
I'm using v1.4.0post0 and can definitely pass NaN to the metrics affected by #2198 and they work as expected when the value of `average` is either `micro` or `none`, but return NaN in the case of `macro`, so I'm computing it manually from the classwise results.

**Comment by DimitrisMantas** [comment]
I wrote this implementation of accuracy:

```python
@verify(UNIQUE)
class MetricReduction(StrEnum):
    MACRO = auto()
    MICRO = auto()
    NONE = auto()


class Accuracy(torchmetrics.classification.MulticlassStatScores):
    def __init__(
        self,
        num_classes: int,
        reduction: MetricReduction,
        include_background: bool = True,
    ):
        super().__init__(
            num_classes=num_classes,
            average="none",
            multidim_average="samplewise",
            ignore_index=None if include_background else 0,
        )

        self.reduction = reduction
        self.include_background = include_background

    def compute(self) -> torch.Tensor:
        """Compute accuracy based on inputs passed in to ``update`` previously."""
        tp, fp, tn, fn = self._final_state()

        if self.reduction == MetricReduction.MICRO:
            # Reduce across the class dimension.
            tp = tp.sum(dim=1)
            fn = fn.sum(dim=1)

        score = torchmetrics.utilities.compute._safe_divide(
            tp, tp + fn, zero_division=torch.nan
        )
        # Reduce across the sample dimension.
        score = score.nanmean(dim=0)

        if self.reduction == MetricReduction.MACRO:
            return score.nanmean()

        return score

```
which appears to be giving me the correct results for this test case, at least:


```python
y_pred=torch.tensor([
    [0,1,2,3], 
    [0,1,2,3],
])

y_true=torch.tensor([
    [0,0,2,3],# Batch 1: 1st Class is missing.
    [0,1,2,3],
])
```
In particular, if we compute samplewise scores and then average them accroe should be:


I wrote this implementation of accuracy because I wasn’t sure what TorchMetrics was doing. It’s rudimentary in comparison but fits my purposes well.
My version computes class- and sample-wise statistics and then reduces them according to the value of `reduction`: `MACRO`, `MICRO` or `NONE`. This parameter works the same way as the existing `average`. Sample reduction is performed by computing the accuracy individually for each item in the given batch before averaging it. This allows the metric to account for absent classes by assigning them a score of NaN, which may be propagated up to the global output if a particular class is entirely absent. The ignored class works is effectively marked as absent, so there’s no need to handle it differently. 

```python
@verify(UNIQUE)
class MetricReduction(StrEnum):
    MACRO = auto()
    MICRO = auto()
    NONE = auto()


class Accuracy(torchmetrics.classification.MulticlassStatScores):
    def __init__(
        self,
        num_classes: int,
        reduction: MetricReduction,
        include_background: bool = True,
    ):
        super().__init__(
            num_classes=num_classes,
            average="none",
            multidim_average="samplewise",
            ignore_index=None if include_background else 0,
        )

        self.reduction = reduction
        self.include_background = include_background

    def compute(self) -> torch.Tensor:
        """Compute accuracy based on inputs passed in to ``update`` previously."""
        tp, fp, tn, fn = self._final_state()

        if self.reduction == MetricReduction.MICRO:
            # Reduce across the class dimension.
            tp = tp.sum(dim=1)
            fn = fn.sum(dim=1)

        score = torchmetrics.utilities.compute._safe_divide(
            tp, tp + fn, zero_division=torch.nan
        )
        # Reduce across the sample dimension.
        score = score.nanmean(dim=0)

        if self.reduction == MetricReduction.MACRO:
            return score.nanmean()

        return score

```

This appears to be giving me correct results for this test case, at least:

```python
y_pred=torch.tensor([
    [0,1,2,3], 
    [0,1,2,3],
])

y_true=torch.tensor([
    [0,0,2,3],# Batch 1: 1st Class is missing.
    [0,1,2,3],
])
```
In particular, if we compute scores separately for each class and sample, and then average them across the batch, we should get:

|                        | `reduction` | `MACRO`  | `MICRO`  | `NONE`                             |
|------------------------|---------------|--------|--------|----------------------------------|
| `include_background` | `True`        | 0.9375 | 0.8750 | [0.7500, 1.0000, 1.0000, 1.0000] |
| `include_background` | `False`       | 1      | 1      | [NaN, 1.0000, 1.0000, 1.0000]      |

Other than assigning 0 instead of NaN to the score of the first class when it is ignored, TorchMetrics agrees with me with the exception of the background score when it is included, which is given as 0.6667 instead of 0.75. I think that this is too strict because the background score in the first sample is 0.5 and 1 in the second, so it would make sense for the global score to be the average.
